### ISSUE TITLE
XSS in p:textEditor

### ISSUE THREAD

**Issue by cnsgithub** [primefaces/primefaces#3214]
## 1) Environment
- PrimeFaces version: primefaces-6.2.RC2-SNAPSHOT
- Application server + version: Wildfly 11
- Affected browsers: all

## 2) Expected behavior
There is a limited set of HTML tags quilljs is able to generate via toolbar. However, it is possible to manipulate the DOM on the client side and submit a tampered version to the server which in turn should sanitize the HTML before persisting or sendig it back to the client.

## 3) Actual behavior
It is possible to inject arbitrary HTML/JS code in DOM. The code could then be persisted to the database and reflected to clients trying to edit using p:textEditor or getting the text rendered through any component that uses `escape="false"`.

## 4) Steps to reproduce
You can test the reflective XSS on your showcase site.
1. Go to https://www.primefaces.org/showcase/ui/input/textEditor.xhtml
2. Open developer tools (F12) and enter in the console: `$('.ql-container').next('input:hidden').val("<script>alert('oops');</script>");`
3. Press the submit button. An alert will be shown after returning from server side.

## 5) Sample XHTML
n/a

## 6) Sample bean
n/a

## 7) Mitigation
This time it is a bit more difficult to fix the issue, because we cannot simply escape all HTML/JS. However, what we can do is to sanitize user HTML input - depending on the quilljs features turned on and thus are available via the toolbar. I.e., if images are not allowed, we should sanitize in a way, that all image tags are dropped and so on. And once again, OWASP is your best friend: https://github.com/OWASP/java-html-sanitizer Btw, I am not a member of OWASP... ;-) I have used it successfully as a workaround for this PrimeFaces issue.

**Comment by melloware** [comment]
Isn't that exactly why escape="false" is offered?  You are turning off the escaping because you know you want to render HTML?  I use it all the time in outputText that I want to format.

**Comment by cnsgithub** [comment]
I have to disagree.

Would you also say it's okay if you had defined a `p:spinner` component with `min="0"` and `max="10"`, but JSF accepted -1, i.e. a value that is out of the bounds?

You can compare this to `p:textEditor`: the developer decides to use it in the assumption the component only allows basic HTML like styling, fonts and images. But unfortunately there is the bad guy that submits arbitrary HTML and the server (PrimeFaces) accepts it without validating.

That's definetely not okay.

**Comment by martin654** [comment]
Hi!
I will share my experience with escape="false". From my point of view it is a security risk using it without thinking about consequences. In our app we used it only for output, but it was always **sanitized** using org.jsoup.Jsoup.clean method. The example @cnsgithub provided definitely is a security risk from my point of view.
Cheers!

**Comment by melloware** [comment]
I see what you mean.  I thought you were saying to auto-cleanse the WYSIWYG entry for the customer.  There may be some use cases where the end user wants to allow JavaScript typed in there.  Maybe they are using this to dynamically edit some web pages that get stored in the DB and then rendered upon command.  So as long as you leave the ability to turn the sanitize completely OFF if the user chooses to.

What makes this component different than say a Spinner validation is that you don't know all the use cases for how someone will use it.  For example WordPress lets you enter HTML and Javascript because that is stored in the database and rendered to the end user that way.    Wordpress understands the whole point of giving you an editor is that you are editing web pages Javascript and all.

**Comment by cnsgithub** [comment]
I am not sure if I got it correctly. Where can javascript be entered in quilljs, p:textEditor resp.? Just can't find an official way to do so.

**Comment by melloware** [comment]
@cnsgithub Maybe you are misunderstanding what I am saying?  Right now do the following:

1. Go to https://www.primefaces.org/showcase/ui/input/textEditor.xhtml
2. Right in the WYSIWYG editor on the screen type `<script>alert('oops');</script>`

You will see the Javascript alert and I **agree** that is bad.

What I am saying is what if I am using the TextEditor component in my webapp to allow admins to edit some text that I want to store in a database that happens to be in the language of Javascript?  So I am using the editor to truly edit Javascript snippets that I store in the database.  I just want to make sure that whatever solution is presented allows this scenario to still be viable and that by cleansing any JS or HTML code you don't actually affect the text that was entered by the user.

**Comment by jxmai** [comment]
Hmm, I wasn't aware of this issue. XSS is doable because quill itself does not sanitize the HTML output. 
A comment which can contain an execution JS on other machines is a bad news ...

Anyway, The expected behavior should be the same as the original editor:

https://www.primefaces.org/showcase/ui/input/editor.xhtml

![image](https://user-images.githubusercontent.com/10467831/35772018-6c118c48-08fc-11e8-9a13-c6c76d2c91ae.png)

---

Someone discussed this problem on quill site. Let's see if those info can help to give a clue.

https://github.com/quilljs/quill/issues/510

https://github.com/quilljs/quill/pull/514

**Comment by cnsgithub** [comment]
Ok, indeed I wasn't aware that it is even simpler and can be done just by copy&paste...

If you really think it could be a valid use case, then we should definetely provide an opt-in attribute for this.

And yes, the old p:editor does not have this copy&paste 'feature'.

**Comment by cnsgithub** [comment]
Pull request: https://github.com/primefaces/primefaces/pull/3221

**Comment by prakhartkg** [comment]
Hi,

Can anyone give me any solution on this I am using primeface texteditor and while putting 
"><img src=x onerror=confirm('hello')>
its executing because I am using h:output text with escap=false

**Comment by cnsgithub** [comment]
@prakhartkg Use e.g. https://github.com/OWASP/java-html-sanitizer by your own as long as my pull request isn't merged.

**Comment by tandraschko** [comment]
We can't use the PR as it depends on Guava, which is to big to be shaded.
Hoping for another HTML Sanitizer PR ;)

**Comment by cnsgithub** [comment]
That's a pity. :cry:  But thanks for giving it a try.

I think other HTML sanitizers are rare...

Let's fork OWASP/java-html-sanitizer and do it without guava just to fit PrimeTek's no-third-party-deps policy. :wink: I just checked the occurrences of guava imports: https://github.com/OWASP/java-html-sanitizer/search?q=in%3Afile+%22import+com.google%22&unscoped_q=in%3Afile+%22import+com.google%22 That's too many, sorry. And it would of course prevent us from version upgrades.

Just an idea, couldn't we just postpone this security _feature_ to 6.3 or 7.0 and add a compile-time dependency to OWASP/java-html-sanitizer only, mentioning it in the user guide as already done for other deps?

![image](https://user-images.githubusercontent.com/13161711/46848701-9f64eb80-cdeb-11e8-865a-9caae24b7010.png)

**Comment by tandraschko** [comment]
the problem is that guava is really a fat boy and adding it as a dependeny could also cause conflicts when other libraries use guava, too.
Same as with ASM, therefore everyone is shading it.

**Comment by cnsgithub** [comment]
Normally this is where jigsaw or wildfly modules or things like that come into play?

**Comment by tandraschko** [comment]
yup

**Comment by cnsgithub** [comment]
Let's fork OWASP/java-html-sanitizer, shade guava there and release a all-in-one version? Or ask them to do so?

**Comment by cnsgithub** [comment]
Asked them: https://github.com/OWASP/java-html-sanitizer/issues/157

**Comment by tandraschko** [comment]
is it the right project? 
seems like no owasp project uses guava, just java-html-sanitizer

**Comment by cnsgithub** [comment]
Fuck, grrrr

**Comment by cnsgithub** [comment]
Just in case someone decides to revive PR #3221, please don't forget to integrate `rel="noopener noreferrer"` as denoted in:
- https://github.com/OWASP/java-html-sanitizer/issues/147
- https://github.com/primefaces/primefaces/issues/4154

**Comment by mertsincan** [comment]
Thanks a lot, @cnsgithub ;)
We decided to add this fix. Github files seem confused. So I've added it manually. Could you please check it?
Many thanks.

**Comment by melloware** [comment]
So does this mean PF brings in the Sanitizer and Guava no matter what now?

**Comment by tandraschko** [comment]
yes
if someone uses textEditor, he must add the sanitizer + guava dependency
we will not shade it until the sanitizer removes guava

**Comment by melloware** [comment]
OK then doesn't the dependency need to be marked with optional or provided and with newer 20181114.1 version?

```xml
        <dependency>
            <groupId>com.googlecode.owasp-java-html-sanitizer</groupId>
            <artifactId>owasp-java-html-sanitizer</artifactId>
            <version>20181114.1</version>
            <optional>true</optional>
        </dependency> 
```

**Comment by melloware** [comment]
OK just so we are all on the same page.  I just tried to run the PF Extensions showcase and got this error parsing faces-config.xml.

> Caused by: com.sun.faces.config.ConfigurationException:
>   Source Document: jar:file:/C:/Users/elefkof/.m2/repository/org/primefaces/primefaces/6.3-SNAPSHOT/primefaces-6.3-SNAPSHOT.jar!/META-INF/faces-config.xml
>   Cause: Class 'org.primefaces.component.texteditor.TextEditorRenderer' is missing a runtime dependency: java.lang.NoClassDefFoundError: org/owasp/html/HtmlPolicyBuilder

So to me it appears this dependency is NOT optional and is now a required compile time dependency.  I am OK with that but then we need to remove the "provided" marker and let everyone know that HTML sanitzier and Guava are now coming with PrimeFaces as transitive dependencies.

I notice it was added to the PF Showcase to make it work.  

```xml
        <!-- HTML sanitizer for the textEditor -->
        <dependency>
            <groupId>com.googlecode.owasp-java-html-sanitizer</groupId>
            <artifactId>owasp-java-html-sanitizer</artifactId>
            <version>20181114.1</version>
        </dependency>
```

So basically any PF app cannot even start up without this dependency?

**Comment by tandraschko** [comment]
i will fix this to make it optional

still bit unhappy with everyting
adding 2,5mb guava for a 200kb lib is bullshit... -> https://twitter.com/olivergierke/status/1065195542916067328

**Comment by cnsgithub** [comment]
@mertsincan I'm happy to read that! :beer:

**Comment by tandraschko** [comment]
@melloware could you retest? (If possible please also test the warning log, no time left for me today)

**Comment by melloware** [comment]
@tandraschko Looks great I was able to start the PE Showcase without the HTML sanitizer dependency.

**Comment by melloware** [comment]
For 7.1 a new attribute secure="true" (default) has been added which requires this library by default.  You will only be able to use the TextEditor without it by opting-out with secure="false".

**Comment by claudioweiler** [comment]
Hi @melloware, @tandraschko 

Just found this one. I think that security constraint `rel='noopener noreferrer'` is not implemented in links sanitizing, as suggested in https://github.com/primefaces/primefaces/issues/3214#issuecomment-432567420 and https://github.com/primefaces/primefaces/commit/822e4cd6f5fa8585fc4ba4133d6b925422751c49#r31402159.

Am I missing something?

**Comment by melloware** [comment]
@claudioweiler feel free to submit a PR.  It looks like its configured correctly to me?

**Comment by claudioweiler** [comment]
Maybe I'm wrong!

What I can find is:
- PF has this https://github.com/primefaces/primefaces/blob/30ef1939c89fb38641cbb78a431345264d0831f6/primefaces/src/main/java/org/primefaces/util/HtmlSanitizer.java#L41
- It's based on Sanitizers.LINK that uses `rel='nofollow'` as default: https://github.com/OWASP/java-html-sanitizer/blob/37569793a12238424ebd1c2ccfc7db011b327088/src/main/java/org/owasp/html/Sanitizers.java#L77

Maybe I'm missing something? Maybe it's not expected behavior?

**Comment by melloware** [comment]
Are you sure you have `secure="true"` set on your TextEditor that is what turns the sanitizer on.

**Comment by claudioweiler** [comment]
Sorry @melloware , I didn't mean to be rude, sorry my poor English!

`secure` is `true` by default, right? (https://primefaces.github.io/primefaces/12_0_0/#/components/texteditor)

I don't have `secure=false`, but I will try with explicit `true`.

But, to be clear, it's expected behavior that `noopener noreferrer` appears on `rel` attribute of links for `target=_blank`?

**Comment by melloware** [comment]
i think that is the intent.  So you may have to debug it! And the OWASP Sanitizer is in your classpath?

**Comment by claudioweiler** [comment]
> And the OWASP Sanitizer is in your classpath?

I think we will never understand each other! Sorry!
